documentation/porting_guidelines.md

```diff
@@ -0,0 +1,391 @@
+	.nintr = main_fiq,
+	.cpu_on = cpu_on_handler,
+	.cpu_off = pm_do_nothing,
+	.cpu_suspend = pm_do_nothing,
+	.cpu_resume = pm_do_nothing,
+	.system_off = pm_do_nothing,
+	.system_reset = pm_do_nothing,
+};
+
+static struct serial8250_uart_data console_data __early_bss;
+
+const struct thread_handlers *generic_boot_get_handlers(void)
+{
+	return &handlers;
+}
+
+void console_init(void)
+{
+	serial8250_uart_init(&console_data, CONSOLE_UART_BASE,
+			     CONSOLE_UART_CLK_IN_HZ, CONSOLE_BAUDRATE);
+	register_serial_console(&console_data.chip);
+}
+```
+
+##### platform_config.h
+This is where the major difference will be seen between different platforms,
+since this is where you do the memory configuration, define base addresses etc.
+We're going to list a few here, but it probably makes more sense to have a look
+at the already existing `platform_config.h` files for the other platforms. Our
+fictive gendev could look like this:
+
+```c
+#ifndef PLATFORM_CONFIG_H
+#define PLATFORM_CONFIG_H
+
+/* Make stacks aligned to data cache line length */
+#define STACK_ALIGNMENT		64
+
+#ifdef ARM64
+#ifdef CFG_WITH_PAGER
+#error "Pager not supported for ARM64"
+#endif
+#endif /* ARM64 */
+
+/* 16550 UART */
+#define CONSOLE_UART_BASE	0xcafebabe /* UART0 */
+#define CONSOLE_BAUDRATE	115200
+#define CONSOLE_UART_CLK_IN_HZ	19200000
+
+#define DRAM0_BASE		0x00000000
+#define DRAM0_SIZE		0x40000000
+
+/* Below ARM-TF */
+#define CFG_SHMEM_START		(0x08000000)
+#define CFG_SHMEM_SIZE		(4 * 1024 * 1024)
+
+/* If your device has SRAM */
+#define TZSRAM_BASE		0x3F000000
+#define TZSRAM_SIZE		(200 * 1024)
+
+/* Otherwise or in addition, use DDR */
+#define TZDRAM_BASE		(CFG_SHMEM_START + CFG_SHMEM_SIZE)
+#define TZDRAM_SIZE		(32 * 1024 * 1024)
+
+#define CFG_TEE_CORE_NB_CORE	4
+
+#define CFG_TEE_RAM_VA_SIZE	(4 * 1024 * 1024)
+
+#define CFG_TEE_LOAD_ADDR	(TZDRAM_BASE + 0x20000)
+
+#define CFG_TEE_RAM_PH_SIZE	CFG_TEE_RAM_VA_SIZE
+#define CFG_TEE_RAM_START	TZDRAM_BASE
+
+#define CFG_TA_RAM_START	ROUNDUP((TZDRAM_BASE + CFG_TEE_RAM_VA_SIZE), \
+					CORE_MMU_DEVICE_SIZE)
+#define CFG_TA_RAM_SIZE        (16 * 1024 * 1024)
+
+#define DEVICE0_BASE		ROUNDDOWN(CONSOLE_UART_BASE, \
+					  CORE_MMU_DEVICE_SIZE)
+#define DEVICE0_PA_BASE         DEVICE0_BASE
+#define DEVICE0_SIZE		CORE_MMU_DEVICE_SIZE
+#define DEVICE0_TYPE		MEM_AREA_IO_NSEC
+
+#endif /* PLATFORM_CONFIG_H */
+```
+This is minimal amount of information in the `platform_config.h` file. I.e, the
+memory layout for on-chip and external RAM. Note that parts of the DDR typically
+will need to be shared with normal world, so there is need for some kind of
+memory firewall for this (more about that further down). As you can see we have
+also added the UART configuration here, i.e., the `DEVICE0_xyz` part.
+
+### 2.2 Devices officially in OP-TEE?
+We do encourage everyone to submit their board support to the OP-TEE project
+itself, so it becomes part of the official releases etc. If you intend to do so,
```

Comment:
Could replace `etc` with `and will be maintain by the OP-TEE community as the OP-TEE frameworks and internal subsystems evolve and mature.` or something like this.